### PR TITLE
Tweet trending news posts

### PR DESCRIPTION
**Pull request by anthotsang**
https://app.asana.com/0/902343675126966/1112458815125768

Actual tweeting is disabled in favour of slack messages at this phase to test on production data.

### CODE REVIEW

File: app/services/news_services/tweet_trending_news_item.rb [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+      }
+      if response_hash['twitter'].present?
+        metadata[:tweet_id] = response_hash['twitter'].try(:id)
+      end
+      if @test_run
+        metadata[:test_run] = true
+      end
+      tweet_body = response_hash['twitter'].present? ? response_hash['twitter'].try(:text) : response_hash['body']
+
+      NewsTweet.create!(news_item_id: @news_item.id,
```
Comment (hanchang) [inline app/services/news_services/tweet_trending_news_item.rb]: This confused me for a sec, might be better formatted as:

```
    NewsTweet.create!(
        news_item_id: @news_item.id,
        tweet_body: tweet_body,
        metadata: metadata,
    )
```

File: test/integration/news_render_test.rb [hunk 2]
```diff
@@ -35,7 +35,7 @@ class NewsRenderTest < ApplicationIntegrationTest
 
   test "show renders news items" do
     news_item = @coins.first.news_items.published.first
-    news_item_slug = news_item.title.parameterize
+    news_item_slug = news_item.slug
```
Comment (hanchang) [inline test/integration/news_render_test.rb]: 👍

File: app/services/news_services/tweet_trending_news_item.rb [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    end
+
+    private
+
+    # Up to 280 characters; links count as 23 characters
+    # Include up to `max_coin_tag_count` coin tags
+    def build_tweet
+      link = build_url news_item_path(id: @news_item.id, slug: @news_item.slug)
+      coin_tags_sorted_by_rank = @news_item.coins.pluck(:ranking, :symbol)
+        .sort_by { |coin| coin[0] || Float::INFINITY }
```
Comment (hanchang) [inline app/services/news_services/tweet_trending_news_item.rb]: What's the `INFINITY` for?

File: app/services/news_services/tweet_trending_news_item.rb [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        .map { |coin| coin[1] }
+        .uniq
+      coin_tag_text = coin_tags_sorted_by_rank.slice(0, @max_coin_tag_count)
+        .map { |coin_tag| '$' + coin_tag }
+        .join(' ')
+
+      # constant ints: max length, url length, spaces
```
Comment (hanchang) [inline app/services/news_services/tweet_trending_news_item.rb]: 🙏

File: app/services/news_services/tweet_trending_news_item.rb [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    def initialize(min_tweet_interval: 4.hours, max_tweets_per_day: 4, min_votes_required: 10, leading_text: "", max_coin_tag_count: 3, test_run: false, store_test_run: false)
+      @min_tweet_interval = min_tweet_interval
+      @max_tweets_per_day = max_tweets_per_day
+      @min_votes_required = min_votes_required
+      @leading_text = leading_text
+      @max_coin_tag_count = max_coin_tag_count
+      @test_run = test_run
+      @store_test_run = store_test_run
+
+      @slack_channel = '#' + ENV.fetch('SLACK_CHANNEL_TRENDING_NEWS').gsub('#', '')
```
Comment (hanchang) [inline app/services/news_services/tweet_trending_news_item.rb]: I'm probably missing something but what's the purpose for prepending `#` and then `gsub`ing it back out?

Might be helpful to leave a code comment!

File: app/services/news_services/tweet_trending_news_item.rb (hunk 3 above)
Comment (anthotsang) [reply app/services/news_services/tweet_trending_news_item.rb]: It's possible the rank could be nil. Want to make sure those get placed at the end.

File: app/services/news_services/tweet_trending_news_item.rb (hunk 5 above)
Comment (anthotsang) [reply app/services/news_services/tweet_trending_news_item.rb]: Pretty sure we can't use hash tags with dot-env (it'll just end up as a comment in the .env file), but it'll probably work with env vars on production. Just wanted to not have to worry about whether or not to hashtag channel names.

**Pull request merged**